### ISSUE TITLE
Extracting Nethack's rendering and commands given

### ISSUE THREAD

**Issue by Sun-2** [NetHack/NetHack#395]
I'd like to write my own frontend for Nethack.

Given I'm not much of a C programmer, I'd like to keep my interaction with C to minimum. My first thought is to find the chokepoints where Nethack gathers keyboard input, and where it outputs to screen. There, I'd set up a socket server to communicate with an in-browser interface where I'd do all the parsing.

If you'd be as kind, please tell me which code should I look into, if I'd like to familiarize myself with how Nethack:
- reads input from the user
- renders the current state of the game

**Comment by ostrosablin** [comment]
I've already considered doing something like that. My particular reasons were to try and integrate NetHack into Minetest (3D Minecraft-styled block sandbox) and use it as GUI for NetHack.

Eventually, I didn't get to actually implement this project. But I've still investigated how to do this (which is pretty close to what you're trying to achieve - get NetHack windowport programming out of C plane, and ultimately - allow handling NetHack's input and output completely outside of the nethack process).

To do that, essentially a new windowport (that's "frontend" in nethack terminology) needs to be implemented (in C). For inter-process communications it should either use pipes (what I originally thought to do), but socket would work fine, too. Windowport should somehow serialize and send over the wire "rendering" calls and other useful information that passes through windowport (menus, etc), while listening for any incoming "keyboard" commands. On receiving side, rendering information could be extracted and used to draw game state in any desirable form (tty-like character graphics, 2D tiles or even 3D).

I think, the best way would be to produce JSON output, describing the glyphs printed, their positions and colors, because JSON is wide used and most modern languages have built-in support for JSON, or 3rd party JSON libraries exist. Basically, all rendering info (function's arguments, etc) could be included into JSON, producing something like JSON RPC protocol. Same could be used for input - client could encapsulate keypresses into JSON messages and send them to nethack process, where they would be deserialized and processed instead of waiting for real keypress.

Of course, you will have to implement actual rendering from scratch on your side, because you can't really get a full state of screen at once. You will only receive information, telling you locations of different objects, that need to be drawn somehow.

**Comment by nhmall** [comment]
I would suggest reading doc/window.doc first.

NetHack uses a set of function pointers to communicate with a platform/windowport, and those are adjusted/updated at runtime reflective of the windowport chosen at the time. Even if you were going to do what @tmp6154 suggested, the least intrusive way to integrate with NetHack is through that interface. Each different existing windowport defined the same set of functions and placed pointers to those functions in a 
    **_struct window_procs_** 
structure.  It's then up to you to add your functionality to the routines.

[skeleton.c.txt](https://github.com/NetHack/NetHack/files/5319298/skeleton.c.txt)

**Comment by ostrosablin** [comment]
@nhmall Yes, that was actually the plan - to implement a complete and compliant (non-intrusive) windowport dedicated to communication with external programs and producing the machine-readable output that would be easy to parse and which would be usable with pretty much any language. And I planned to use function pointer interface that other windowports are using to intergrate.

**Comment by paxed** [comment]
Look up NetHack-el - it created a new windowport outputting lisp, so you could use emacs to play NetHack.

**Comment by moon-chilled** [comment]
Two other good places to look are slashem and nethack4.

Slashem's 'proxy' windowport was essentially a network protocol for the game.  Slashem's core code is fairly close to nethack's; the windowport could probably be ported to 3.6 without too much trouble.

Nethack4's code is much different to nethack's, but it also implements a network protocol, based on JSON iirc.  Ais can probably talk more about that.

**Comment by actual-nh** [comment]
You may also find [NLE](https://github.com/facebookresearch/nle), the Nethack Learning Environment, of use; also the libnethack and "shim" work in [Nethack-3.7](https://github.com/facebookresearch/nle/issues/95) may be of interest.

@sun-2: The NLE work interfaces Nethack with Python; the "shim" work does so with Java.

**Comment by nhmall** [comment]
This issue was opened during one of the 3.6 versions of NetHack which are no longer being actively maintained. 

Please try the NetHack 5.0.0 release, which also includes the libnethack and "shim" work mentioned in the thread above.
Thank you.
